File: Yandex/Practicum/Algorightms_and_Data_Structures/8_sprint/F._Частое_слово/solution.cpp
#include <iostream>
#include <string>
#include <unordered_map>


void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(D)
    */
    int n;
    std::scanf("%d", &n);
    std::unordered_map<std::string, int> gistogram;
    int maxFreq = 0;
    std::string maxFreqWord = "";
    for (int i = 0; i < n; ++i) {
        std::string word;
        std::cin >> word;
        ++gistogram[word];
        if (gistogram[word] > maxFreq) {
            maxFreq = gistogram[word];
            maxFreqWord = word;
        } else if (gistogram[word] == maxFreq && word < maxFreqWord) {
            maxFreqWord = word;
        }
    }
    std::cout << maxFreqWord << "\n";
}


int main() {
    solution();
}

File: Yandex/Practicum/Algorightms_and_Data_Structures/8_sprint/F._Частое_слово/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(D)
    """
    n: int = int(sys.stdin.readline().rstrip())
    histogram: dict[str, int] = dict()
    max_freq: int = 0
    max_freq_word: str = ''
    for _ in range(n):
        word: str = sys.stdin.readline().rstrip()
        histogram[word] = histogram.get(word, 0) + 1
        if histogram[word] > max_freq:
            max_freq = histogram[word]
            max_freq_word = word
        elif histogram[word] == max_freq and word < max_freq_word:
            max_freq_word = word
    sys.stdout.write('{}\n'.format(max_freq_word))


if __name__ == '__main__':
    solution()